lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2214,124 +2214,70 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
       if (resolvedFuncDecl) {
-        TC.diagnose(funcSpecifier.Loc.getBaseNameLoc(),
+        TC.diagnose(funcNameLoc,
                     diag::differentiable_attr_ambiguous_function_identifier,
-                    funcSpecifier.Name);
-        return nullptr;
+                    funcName);
+        resolvedFuncDecl = nullptr;
+        break;
       }
-      // Resolve the function declaration.
       resolvedFuncDecl = funcDecl;
     }
-    // If the function declaration could not be resolved, check error flags.
-    if (!resolvedFuncDecl) {
-      if (overloadNotFound) {
-        overloadDiagnostic();
-        return nullptr;
-      }
-      assert(exprIsNotFunction && "Function declaration could not be resolved");
-      TC.diagnose(funcSpecifier.Loc.getBaseNameLoc(),
-                  diag::differentiable_attr_specified_not_function,
-                  funcSpecifier.Name, isPrimal);
-      return nullptr;
-    }
   }
-  // If it's resolved to a type, it's not what we want.
-  else if (isa<TypeExpr>(expr))
-    TC.diagnose(funcSpecifier.Loc.getBaseNameLoc(),
-                diag::differentiable_attr_specified_not_function,
-                funcSpecifier.Name, isPrimal);
-  // If it's directly resolved to a concrete declaration, it must be a free
-  // function in the module context.
-  else if (auto declRefExpr = dyn_cast<DeclRefExpr>(expr)) {
-    auto funcDecl = dyn_cast<FuncDecl>(declRefExpr->getDecl());
-    // If the candidate is not a function, then it's an error.
-    if (!funcDecl) {
-      TC.diagnose(funcSpecifier.Loc.getBaseNameLoc(),
-                  diag::differentiable_attr_specified_not_function,
-                  funcSpecifier.Name, isPrimal);
-      return nullptr;
+  // If function declaration could not be resolved, emit the appropriate
+  // diagnostic.
+  if (!resolvedFuncDecl) {
+    if (results.empty()) {
+      TC.diagnose(funcNameLoc, diag::use_unresolved_identifier, funcName,
+                  funcName.isOperator());
     }
-
-    // If the original and the primal or adjoint have different parents, or
-    // if they both have no type context and are in different modules, then
-    // it's an error.
-    auto inCompatibleContexts = [&](FuncDecl *decl1, FuncDecl *decl2) {
-      if (!decl1->getInnermostTypeContext() &&
-          !decl2->getInnermostTypeContext() &&
-          decl1->getParentModule() == decl2->getParentModule())
-        return true;
-      if (decl1->getParent() == decl2->getParent())
-        return true;
-      return false;
-    };
-
-    if (!inCompatibleContexts(original, funcDecl)) {
-      TC.diagnose(funcSpecifier.Loc.getBaseNameLoc(),
+    else if (wrongTypeContext) {
+      TC.diagnose(funcNameLoc,
                   diag::differentiable_attr_function_not_same_type_context,
-                  funcSpecifier.Name);
-      return nullptr;
+                  funcName);
+    }
+    else if (overloadNotFound) {
+      TC.diagnose(funcNameLoc, overloadDiagnostic);
+    } else {
```

Comment:
Wrap this `else` because you wrapped the previous ones.

Reply:
I actually unwrapped all the `if else`/`else`s because I prefer that.
Things are uniform now.